new inputhandler

DeveloperMarius opened the pull request:
Hello @skipperbent ,

while the project wasn't updated in a while I cloned your project into a private repository to change some things in the InputHandler.
Due to the fact that you now seem to create a new version and include changes I want to ask you if you can apply my changes too?

I also wanted to add that I do not changed the tests, but I think it has to be done due to the fact that I changed some methods.

Let's begin:

I stored the content-type header in the Request object to use it later in the InputHandler. Important is that this parameter has to be initialized before the InputHandler object is created.

I made the main changes in the InputHandler. In the parseInputs() function I automatically load the POST vars if the method is post. Because when the request is not post, all values are stored in the request body.
Lets come to the request body. I added two new parameters to the object. "body_plain" which basically stores the recieved body in plain text and the "body". "body" included all parsed body contents. Now I used the content-type from the request object to check if the body has a json or urlencoded content. In the result I just parsed the contents into InputItems. It is important to use the contains function, becuase you can have a content-type like "application/x-www-form-urlencoded; charset=UTF-8".
Using ajax people can for example in a PATCH request submit a json object via the body or urlencoded contents. So I basically added support for that.
Then I just added the body params to the find() and all() functions.
I created a toValue() functions which simply parses a InputItem or an array of Input Items into a value or array. I used foreach $key => $value to add support of sequential and associative arrays. So if the array is sequential the index is simply used as key. This does not make any difference in handling these arrays:
```
$arr = json_decode($input, true);
$arr2 = array();
foreach($arr as $key => $value){
    $arr2[$key] = $value;
}
echo json_encode($arr2); 

$input = '["a","b","c"]'; -> ["a","b","c"]
$input = '{"A":"a","B":"b","C":"c"}'; -> {"A":"a","B":"b","C":"c"}
```

I think the biggest "change" was that the get(), post(), etc. functions now return the real mixed content and not the InputItem. I think you mostly need to work with the value and not with the InputItem. When you really need the InputItem you can not just use postItem(), getItem(), etc.
This changes also impact the find() and all() function. I do not added a function to return the InputItems for these, but it can be added easily.
I think this changes makes the value() function kind of unnessesary, but I'm not sure.
I removed the repeatedly parsing of the $_GET, $_POST and body in the all function and just accessed the already parsed parameters, because I do not see an advantage in fetching them again.
I'm now also returning null if the parameter is not in the request, but I think you already added that into your 4.3.0.0 version.

The last change I made is with the InputItem and the value. I hated it, that when I submitted data from js via ajax I do get a different type when the data was urlencoded. So I added a parser that checks if the value is a boolean and then parses it. 
Also when I submitted a null value it never came to the request values or was an empty string. For that I just declerated all empty string as null. This is a big change but I think it is very usefull and when you just add a "can be null" to your db or check before inserting if the value is null and create an empty string it realy does the same.

I think it is usefull when you change the value of the InputItems into a mixed type. It allowes us to work with booleans and integers instead of checking every time if it is an boolean `$value === 'true'`. Later on it would be great to add support for integers, but For now I'm interested in hearing what you think.

In the end I just wanted to thank you for this awesome project and really hope, that my changes are interesting for you.

Your sincerely,

Marius K.

skipperbent commented:
Hi Marius.

First of all thanks for the nice words at the ends, believe it or not I've spent so much time on this project and one of the things that made me take a brake from it was the fact that people wanted so much support and help with everything - without even bothering to read the documentation or appreciate the fact that the software is free. So it's nice to hear someone enjoying it and even building more features for it!

I think it's some great changes you've made. I will have to do some testing and write some unit-tests for the new features you've made. If you've refactored too many methods it might not make it until the next release - as it will break and require people to update their code; which is only something that can be done on major releases. But a major release is underway, I want to update the php-version and I've got some ideas for new features that can be added for the next major release.

There's already a todo on my list to refactor the `InputHandler` and I'll make sure to include your changes and thoughts. There's also some great ideas in the issues-section too, but it's hard work to sort through them all.

I will let you know when I've got some more info.

Thanks for your contribution and kind words.

/ SImon

DeveloperMarius commented:
Hello Simon,

thank you, that sounds great!
I already worried about that fact that I changed too mutch with the `get()`, `post()`, etc. functions. I'm looking forward to the next major release.
I'm currently using PHP8 in all of my projects, so it would be great to add support for that too, but it is not that important, because you still can use the `--ignore-platform-reqs` parameter with composer and almost none other project supports it now. An example is the big problem with Imagick...
Looking forward to your updates.

~ Marius

DeveloperMarius commented:
Hey,

I worked over some methods and added support for the tests.
I'm just not confident with the methods 
`public function find(string $index, ...$methods)` and 
`public function findItem(string $index, ...$methods)`
I first wanted to add the parameter `$inputItem` like I added in `public function post(string $index, $defaultValue = null, bool $inputItem = false)`, but this I not possible because of the `...$methods` and I also don't want to add this parameter before methods because I think `$methods` is used more.
When I have in mind what to do against this, I will add the same to `value()`.

I hope that I don't work against something in your concept or some ideas of you. Just tell me and I will ajust my request^^

~ Marius

skipperbent commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -279,47 +335,87 @@ public function exists(string $index, ...$methods): bool
     {
-        $output = $this->originalParams + $this->originalPost + $this->originalFile;
-        $output = (\count($filter) > 0) ? \array_intersect_key($output, \array_flip($filter)) : $output;
+        $output = array_merge($this->get, $this->post, $this->body);
```
`$this->file` is no longer here.

DeveloperMarius replied:
👍🏻

skipperbent commented on tests/Pecee/SimpleRouter/InputHandlerTest.php:
```diff
@@ -52,14 +56,14 @@ public function testPost()
+        $objects = $handler->findItem('names');
 
-        $this->assertInstanceOf(\Pecee\Http\Input\InputItem::class, $objects);
+        //TODO add in PHPUnit 8 $this->assertIsArray($objects);
```
`$this->assertTrue(\is_array($objects));` would do the same

skipperbent commented on tests/tmp/CompiledContainer.php:
```diff
@@ -0,0 +1,9 @@
+<?php
```
What is this :)

DeveloperMarius replied:
I'm not sure really xD I will remove it

skipperbent reviewed (changes requested):
I've added some comments to your code.

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -62,33 +78,46 @@ public function __construct(Request $request)
-                if ($post !== false) {
-                    $this->originalPost += $post;
+        /* Parse body */
+        if (in_array($this->request->getMethod(), ['put', 'patch', 'delete', 'post'], false)) {
```
Sure 👍🏻

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -62,33 +78,46 @@ public function __construct(Request $request)
-                    $this->originalPost += $post;
+        /* Parse body */
+        if (in_array($this->request->getMethod(), ['put', 'patch', 'delete', 'post'], false)) {
+            if(strpos($this->request->getContentType(), 'application/json') !== false){
```
true and you can extend it better 👍🏻

DeveloperMarius replied:
Ah, this is not possible. As I described here:

> It is important to use the contains function, because you can have a content-type like `application/x-www-form-urlencoded; charset=UTF-8`.

skipperbent replied:
Nice spottet! Then my suggestion would not work... I've added your `getContentType` method to the `Request` class. Instead of setting it manually I've added so it uses `setContentType` that parses the content-type and strips the charset part so that the result is always uniform. I've also added some generic `CONTENT_TYPE` constants you can use `Request::CONTENT_TYPE_X_FORM_ENCODED` and so on...

DeveloperMarius replied:
I already added them 👍🏻

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -201,11 +230,12 @@ protected function parseInputItem(array $array): array
             // Handle array input
-            if (\is_array($value) === true) {
-                $value = $this->parseInputItem($value);
+            if (\is_array($value) === false) {
```
I will look over it but I used the same code in my current project and it worked.

skipperbent replied:
I need to create some unit tests for when items are nested arrays, because I got some funky results with the current release ($_POST['myData'] is an array in the above example and i think the `InputHandler` in your code will treat all sub-items as `InputItems` which they aren't).

I could be wrong, that's why the unit-tests makes life easier even though it can be a bit tedious to create.

DeveloperMarius replied:
Ahh, I know what you mean. Yes the sub elements of arrays are threaded as Input Items. It was intentionally but I will change that.

skipperbent replied:
It's the same behavior in the current release. It's not that common to use nested arrays like that, but when you do you get some funky results where indexes are missing etc because everything is wrapped as a `InputItem`. I changed it in `v4-development`.

I'll add some more unit-tests for the `InputHandler` soon, that makes it a little easier when making new features.

DeveloperMarius replied:
👍🏻

DeveloperMarius replied:
You changed the function to the following in v4-development:
 ```
protected function parseInputItem(array $array): array
{
    $list = [];

    foreach ($array as $key => $value) {

        // Handle array input
        if (\is_array($value) === true) {
            $value = $this->parseInputItem($value);
        }

        $list[$key] = new InputItem($key, $value);
    }

    return $list;
}
```
But this will thread sub arrays as InputItems...
if the value is an array it will be changed into an array of InputItems, because of the rekursion and then set as the value of the InputItem.

When you don't want to have array contents wrapped into an InputItem you have to do the following:
```
protected function parseInputItem(array $array): array
{
    $list = [];

    foreach ($array as $key => $value) {
        $list[$key] = new InputItem($key, $value);
    }

    return $list;
}
```
We have mixed InputItem values now. So we can the an array as value.
Do I misunderstand something?

~ Marius

DeveloperMarius replied:
Or do you want to handle array contents also as InputItems? Because if the opposite is the case you do not need the `getValueFromArray` function^^

skipperbent replied:
I added better support nested array (before it would sometimes overwrite the index name on certain nested arrays), but more importantly I fixed `InputHandler::value` method so it ALWAYS returns the original value and not the `InputItem` (that's what getValueFromArray is for). Before it would return an array of InputItems, which is not the original value. 

Maybe it should be renamed to `parseValueFromArray` maybe that would make more sense?

DeveloperMarius replied:
Yes, this would sound better I think.
Ahh, now I got your point 👍🏻

DeveloperMarius replied:
What is if you run `post()` get an `InputItem` and call `getValue()`? Shouldn't we add `parseValueFromArray` to the getValue() function? Wouldn't it be useless then?

skipperbent replied:
You have a good point. You're right - it would make sense 👍

DeveloperMarius replied:
So why should we still parse input array contents?

DeveloperMarius replied:
Ahh, no. Let's still parse it, but on the InputItem you can now call `getValue` (get the normal contents) or `getInputItem()` to get the InputItems if it is an array.

skipperbent replied:
Some values are nested like arrays if people use things like `input name="test[]` or `input name="test[data][]`. It's a little confusion, maybe the right thing to do would be to add some complex unit-tests and see what makes sense. Because some of the things I'm not entirely sure about, but I know it was added for a reason.

I'm planning on adding input-validation support too in the next version.

DeveloperMarius replied:
You always think about form contents but what about ajax contents? I submit data via ajax with nested arrays. You don't do that with x-urlenc or form-data but with json content-type.

Can I just add the input-validation to the new inputhandler? I just started with that

skipperbent replied:
That's a great idea! That gives the control back to the developer, he can chose what output he wants.

DeveloperMarius replied:
Okay 👍🏻

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -215,52 +245,78 @@ protected function parseInputItem(array $array): array
+     */
+    private function toValue($value)
+    {
+        if($value === null)
```
You are right. This was mainly created for ajax submits where I set the param to a specific type in js. But you are right 👍🏻.
Yes we kan create single functions, but what do you think about just providing a function like
```
public function parseValue(){
        if($this->value === 'true')
            return true;
        else if($this->value === 'false')
            return false;
        else if($this->value === '')
            return null;
        else
            return $this->value;
}
```
or do you preferr adding both (single casting and just parsing)?

skipperbent replied:
The developer knows what object to expect, so I don't really think there's a point in parsing the value. Plus the parse could be wrong, floats can easily be mistaken for integers etc.

If we create a method for the most common types, we only need to cast the value in most cases:

```php
function getValueAsInteger(): int {
   return (int)$this->value;
}
```

DeveloperMarius replied:
Okay, sure 👍🏻. But I will add a small validation. Yes the developers know what they want to recieve, but not every developer is validating the incoming data and this could lead to erros and security issues.

skipperbent replied:
I agree, plus warnings could be disabled or not clearly visible.

Maybe throw some kind of `InvalidCastException` etc. That way the developer will know the risks of casing incorrect values when calling the method and have to take that into account in his project. This is very similar to other languages.

```php
if(\is_int($this->value) === false) {
   throw new InvalidCastException('Failed to cast value. The value is not a valid integer.');
}
return (int)$this->value;
```

DeveloperMarius replied:
👍🏻 Great idea

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -215,52 +245,78 @@ protected function parseInputItem(array $array): array
+                $data[$key] = $this->toValue($subitem);
+            }
+            return $data;
+        }else{
```
👍🏻

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -279,47 +335,87 @@ public function exists(string $index, ...$methods): bool
+     * @return mixed
      */
-    public function post(string $index, ?string $defaultValue = null)
+    public function post(string $index, $defaultValue = null, bool $inputItem = false)
```
👍🏻  okay, I wil change it

DeveloperMarius replied:
In the next push I now always return an `InputItem` and when calling `file()` I return an `InputFile`.
How do I handle the `$defaultValue` with `file()`? I currently decided to return `null` and do not give the option of a `$defaultValue`.

skipperbent replied:
I think it's just an `IInputItem` type, so something like `(new InputFile('key'))->setValue($defaultValue);` should do it.

I think that would make this work:

```php
if(null === input()->file('my_file', null)->getValue()) {
   // awesome things happen
}
```

DeveloperMarius replied:
👍🏻

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -62,33 +78,46 @@ public function __construct(Request $request)
+        $this->get = [];
+        $this->originalPost = [];
+        $this->post = [];
+        $this->body_plain = '';
```
👍🏻

skipperbent commented:
Would be nice if you added unit-tests for the new feature you've implemented.

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -62,33 +78,46 @@ public function __construct(Request $request)
+            }else if(strpos($this->request->getContentType(), 'application/x-www-form-urlencoded') !== false){
+                parse_str(file_get_contents('php://input'), $body);
+                $this->originalBody = $body;
+                $this->body = $this->parseInputItem($this->originalBody);
```
Hmm, I can test that but I thought that `PATCH` values for example only show up in the body and not the post.
I will test this and update

skipperbent replied:
`$this->originalBody` already contains `php://input` values.

DeveloperMarius replied:
Yes you are right with that

skipperbent replied:
Wouldn't this do the trick `$this->post += $this->parseInputItem($this->originalBody)`? Why is separate property `body` needed? I'm pretty sure that `php://input` is raw `$_POST` values from other request-types. Maybe I'm mistaken...

DeveloperMarius replied:
I will research a bit and write you then

DeveloperMarius replied:
Okay,
application/x-www-form-urlencoded -> data in $_POST and php://input
multipart/form-data -> data in $_POST (and $_FILES)
application/json -> data in php://input

$_POST cannot parse JSON data. I will write more on it in the next comment after my push.

DeveloperMarius replied:
Do you want to add json data to the post data? I think the name `POST` can lead to misunderstandings here because we mix request type with data store location. I can send a `PATCH` request using `application/x-www-form-urlencoded` and the data is stored in the `$_POST`.

We can still use the variable name `post` but we have to return the data for the following requests:
`POST, PUT, PATCH, DELETE`

What do you think?

skipperbent replied:
I think request-types like PATCH will be treated like POST on the webserver which is why the data can be found in `$_POST` (unless its special request-types like defined in php.net description). You can only have one request-type, so might as well merge the results of `php://input` to the post parameter. I think that's how it is currently.

DeveloperMarius replied:
Okay, I have some ideas for that. You can see later in the push how I changed it.

DeveloperMarius commented:
I will work on the changes and notify you if it's done

skipperbent commented:
I forgot to add great work, you're working like a machine :)

DeveloperMarius commented:
Thank you! I have to work on other projects today so I think that the changes will be made tomorrow.

DeveloperMarius commented:
Hey,

why do you have `$this->method = strtolower($this->inputHandler->value('_method', $this->getHeader('request-method')))` in `Request@____construct`?

DeveloperMarius commented:
So, I pushed my seconds round of changes now.
The RequestMethod have to be set before calling `InputHandler@parseInputs`, to parse the right contents.
For that I added `public function reset(string $method = null)` and `public function __construct(string $method = null)` to set the method in the tests.

Nested InputFiles now have an InputFile containing all other InputFiles in getValue or getInputItems.
I thought about that a lot, but I would change
```
public function getInputItems()
{
    if(is_array($this->getValue())){
        return $this->value;
    }
    return array();
}
```
to 
```
public function getInputItems()
{
    if(is_array($this->getValue())){
        return $this->value;
    }
    return array($this);
}
```
because when I do it like that I can get the current file also when I'm iterating thru the class using `ArrayIterator` or just get the the Item when I thought (what shouldn't happen) it was an array that I recieve. The same with `InputItem`. What do you think?

I added the support to always return an return an instance of `IInputItem` and changed the tests for that.

I thought a lot about that, but I wouldn't store the `$_POST` and parsed body data in seperate arrays. I put them into $data and let the developers access them with `POST, PUT, PATCH and DELETE` and using the functions `post()` and `data()`.
I also still store the `$originalPost` and `$originalBody` and `$originalBodyPlain` fot the `all()` functions, where we do not return InputItems. I wanted to ask you I we can change that behaviour and return InputItems using `all()`? Then I could remove `$origionalBody` and rename `$originalBodyPlain` to `$origionalBody`.

In `find()` I had to move the search for a `file()` before the search for `data()`, because when nothing is found it would return an `InputFile` instead of and `InputItem`.

Then I moved `parseValueFromArray` into the InputItem. You can now get the InputItems (if it is an array) using `getInputItems()` and check if this is an array or if there are `InputItems` using `hasInputItems()`. (Same for files)

Other questions:
In `find()` I don't know what to do with the reuqest method for file. Do you have an idea?
For `switch($this->request->getContentType()){` i added `Request::CONTENT_TYPE_X_FORM_ENCODED and Request::CONTENT_TYPE_FORM_DATA` as default. Should I change that?
I'm also planning to support xml data in the next push. What do you think?

I'm creating a merge for the Input validator this evening or tomorrow morning. I'm from Germany by the way, so timezone `Central European Time (CET)`. What is yours? Then I can plan better when to expect feedback 😄 

~ Marius

DeveloperMarius commented:
Hey @skipperbent,
I know you had a lot of work todo with the release and now the fixes, but it would be very nice if you now can take yourself some time to look info my pull requests. This would mean a lot to me^^

~ Marius

skipperbent commented:
Hi Marius.

You have not been forgotten. I've been really busy today with fixing issues in the project, but I promise you that I'll take a look at it. Your feedback was the reason that so many of the issues got caught today 👍

DeveloperMarius commented:
Yes, I kept you away from my pull requests 😄
No problem, happy to help 👍🏻

DeveloperMarius commented:
Hey @skipperbent,

hope you are doing good?
Did you have time to look over my code?

~ Marius

DeveloperMarius commented on src/Pecee/Http/Input/InputHandler.php:
```diff
@@ -128,10 +149,11 @@ public function parseFiles(array $files, ?string $parentKey = null): array
 
+            /** @var InputFile[] $list */
             if (isset($list[$key]) === true) {
-                $list[$key][] = $files;
+                $list[$key]->addInputFile(array_values($files));
             } else {
-                $list[$key] = $files;
+                $list[$key] = (new InputFile($key))->setValue(array_values($files));
```
We don't have a unittest for this case. I'm not completly sure if it works now after the ca552d5a80e0ccbeb5bcbf7c10dcda459a830acf
@skipperbent

DeveloperMarius commented:
@skipperbent, since you are integrating features by yourself instead of working with me on them, when I already added some, I am some sort of disappointed. I mean you don't have to let me work on your project, but when you want to integrate certain features by yourself, just tell me so and I don't have to work on them or wait for your feedback.
I'm stopping to put my time into this until I get feedback and will integrate my changes in my private fork. I really wanted to use your original project, but I can't wait forever and don't want to work on something when I just always get told that I get feedback soon and nothing happens.

Still hoping to continue working with you in the future, but I won't work like this.

Until then, stay save and healthy,

~ Marius

DeveloperMarius commented:
Hello,

I will close this PR, because we are searching for a better alternative in PR #607.

~ Marius
